bluetooth: controller: Add ISO-AL TX unframed fragmentation

niag-Oticon opened the pull request:
- Implemented ISO-AL TX interface functions for fragmentation of
  unframed PDUs
- Implemented ISO-AL source construct and its creation for an input
  data path

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

mtpr-ot commented:
@cvinayak, please review. We need to get ISOAL TX in to complete the data path. Thanks

mtpr-ot commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -9,27 +9,39 @@
 /* TODO this must be taken from a Kconfig */
 #define ISOAL_SINKS_MAX   (4)
+#define ISOAL_SOURCES_MAX (4)
 
```
Suggest to create Kconfigs now

mtpr-ot commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -9,27 +9,39 @@
 
 /* TODO this must be taken from a Kconfig */
 #define ISOAL_SINKS_MAX   (4)
+#define ISOAL_SOURCES_MAX (4)
```
Suggest to create Kconfigs now

wopu-ot replied:
Come to think of it, we will need a separate Kconfig for this. A device could have one microphone and one speaker, i.e., one source and one sink, but support uplink and downlink in separate CISes. Attempts at guessing how many streams a device needs to support will inevitably go wrong, so this needs to be a separate configuration. For now, leaving it as is is fine by me.

mtpr-ot reviewed (dismissed):
LGTM

wopu-ot commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -212,6 +224,110 @@ isoal_status_t isoal_sink_create(
+	 */
+
+	/* Note: sdu_interval unit is uS, iso_interval is a multiple of 1.25mS */
+	session->pdus_per_sdu = burst_number * (sdu_interval / (iso_interval * 1250));
```
I think this should be something like `(burst_number * sdu_interval) / ((uint32_t)iso_interval * 1250)` to avoid rounding errors.

niag-Oticon replied:
Done.

cvinayak commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -212,6 +224,110 @@ isoal_status_t isoal_sink_create(
+	 */
+
+	/* Note: sdu_interval unit is uS, iso_interval is a multiple of 1.25mS */
+	session->pdus_per_sdu = burst_number * (sdu_interval / (iso_interval * 1250));
```
Use `PERIODIC_INT_UNIT_US` or `CONN_INT_UNIT_US`
```suggestion
	session->pdus_per_sdu = burst_number * (sdu_interval / (iso_interval * PERIODIC_INT_UNIT_US));
```

niag-Oticon replied:
Done.

niag-Oticon replied:
@cvinayak Would be great if I could have a review of the update.
Thanks.

cvinayak commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -775,3 +941,313 @@ isoal_status_t isoal_rx_pdu_recombine(isoal_sink_handle_t sink_hdl,
+static isoal_status_t isoal_tx_unframed_produce(struct isoal_source *source,
+						const struct isoal_sdu_tx *tx_sdu)
+{
+	struct isoal_source_session *session;
+	struct isoal_pdu_production *pp;
+	isoal_sdu_len_t packet_available;
+	const uint8_t   *sdu_payload;
+	isoal_status_t  err;
+	bool zero_length_sdu;
+	bool padding_pdu;
+	uint8_t ll_id;
```
```suggestion
	struct isoal_source_session *session;
	isoal_sdu_len_t packet_available;
	struct isoal_pdu_production *pp;
	const uint8_t *sdu_payload;
	bool zero_length_sdu;
	isoal_status_t err;
	bool padding_pdu;
	uint8_t ll_id;
```

niag-Oticon replied:
Done.

cvinayak reviewed (changes requested):
Quick pass through, minor changes requested, otherwise looks good to me.

niag-Oticon commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -775,3 +941,313 @@ isoal_status_t isoal_rx_pdu_recombine(isoal_sink_handle_t sink_hdl,
+	isoal_status_t  err;
+	bool zero_length_sdu;
+	bool padding_pdu;
+	uint8_t ll_id;
```
Done.

wopu-ot commented on tests/bluetooth/ctrl_isoal/CMakeLists.txt:
```diff
@@ -15,4 +15,6 @@ zephyr_library_include_directories(
 
 FILE(GLOB app_sources src/*.c)
 
+add_compile_definitions(CONFIG_BT_CTLR_ISOAL_SINKS=4 CONFIG_BT_CTLR_ISOAL_SOURCES=4)
```
I believe this should go into prj.conf

mtpr-ot replied:
This is my suggestion. The problem with using `prj.conf` is that you may end up activating a lot more code than you want due to dependencies. A unit test needs to be able to have full control of its dependencies, and for that reason I suggested to allow this kind of override.

wopu-ot replied:
Thank you for the explanation. I found some precedence in other CMakeLists.txt for test cases, so I will approve.

ppryga replied:
It looks that this is another approach to get full control over configuration by tests. 
Here you can see another approach that was used for LLCPs: https://github.com/zephyrproject-rtos/zephyr/pull/40073#issuecomment-1044510294.

@tejlmand have you got any suggestion how to follow with this king of issues in general?

tejlmand replied:
> The problem with using `prj.conf` is that you may end up activating a lot more code than you want due to dependencies.

No, what you have to do in this case is to take the needed Kconfig settings into a local `Kconfig` file to make them locally available and thus removing all dependencies, like this:
```
# tests/bluetooth/ctrl_isoal/Kconfig

config BT_CTLR_ISOAL_SINKS
        int "Number of Isochronous Adaptation Layer sinks (for unittest)"

config BT_CTLR_ISOAL_SOURCES
        int "Number of Isochronous Adaptation Layer sinks (for unittest)"

source "Kconfig.zephyr"
```

This will then allow you to set the test values in the standard way using `prj.conf`, like this:
```
# prj.conf
CONFIG_BT_CTLR_ISOAL_SINKS=4
CONFIG_BT_CTLR_ISOAL_SOURCES=4
....
```

A little more work, but it keeps the Kconfig infrastructure intact and all settings correctly available to CMake, autoconf.h and any python script relying on those settings.
Also developers will know where to look for configuration settings.

Finally, your current approach only sets it as compile flag.
This might be enough in this particular case, but is by-passing the normal flow and creating a mess wrt. the design.

niag-Oticon replied:
Thanks. I did not know that the Kconfig could be locally overridden. That's a very nice solution.

tejlmand replied:
thanks for applying the proposal.

wopu-ot reviewed (changes requested):
I think the new Kconfigs should be set in prj.conf for the respective test case, the rest LGTM

tejlmand reviewed (changes requested):
Please use existing Kconfig infrastructure.

niag-Oticon committed:
bluetooth: controller: Add ISO-AL TX unframed fragmentation

- Implemented ISO-AL TX interface functions for fragmentation of
  unframed PDUs
- Implemented ISO-AL source construct and its creation for an input
  data path

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

tejlmand reviewed (approved):
lgtm

Thalley commented:
@wopu-ot @niag-Oticon @mtpr-ot This is still breaking CI when building with ISO enabled.

niag-Oticon commented:
@Thalley Thanks for the info. The fix is in PR https://github.com/zephyrproject-rtos/zephyr/pull/44174.
It's still pending review.
